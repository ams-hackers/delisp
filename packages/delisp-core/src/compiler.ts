import { ASExpr, ASExprSymbol, ASExprList } from "./syntax";
import * as recast from "recast";

const debug = require("debug")("delisp:compiler");

type JSAST = object;
interface Environment {
  [symbol: string]: { symbol: ASExprSymbol; jsVar: string };
}

function parseLambdaList(x: ASExpr): ASExprSymbol[] {
  if (x.type !== "list")
    throw new Error("First argument of lambda should be a list");

  if (!x.elements.every(param => param.type === "symbol"))
    throw new Error("First argument of lambda should be a list of symbols");

  return x.elements as ASExprSymbol[];
}

function compileLambda(args: ASExpr[], env: Environment): JSAST {
  if (args.length !== 2) throw new Error("Lambda needs exactly 2 arguments");

  const params = parseLambdaList(args[0]);

  const newEnv = params.reduce(
    (e, param, ix) => ({
      ...e,
      [param.name]: { symbol: param, jsVar: `p${ix}` }
    }),
    env
  );

  return {
    type: "ArrowFunctionExpression",
    params: params.map(symbol => ({
      type: "Identifier",
      name: newEnv[symbol.name].jsVar
    })),
    body: compile(args[1], newEnv)
  };
}

function compileDefinition(args: ASExpr[], env: Environment): JSAST {
  if (args.length !== 2) throw new Error("Define needs exactly 2 arguments");

  const variable = args[0];
  if (variable.type !== "symbol")
    throw new Error("First argument of define should be a symbol");

  return {
    type: "AssignmentExpression",
    operator: "=",
    left: {
      type: "MemberExpression",
      computed: true,
      object: {
        type: "Identifier",
        name: "env"
      },
      property: {
        type: "Literal",
        value: variable.name
      }
    },
    right: compile(args[1], env)
  };
}

function compileList(fn: ASExpr, args: ASExpr[], env: Environment): JSAST {
  if (fn.type === "symbol") {
    switch (fn.name) {
      case "lambda":
        return compileLambda(args, env);
    }
  }

  return {
    type: "CallExpression",
    callee: compile(fn, env),
    arguments: args.map(arg => compile(arg, env))
  };
}

function compileVariable(symbol: ASExprSymbol, env: Environment): JSAST {
  if (symbol.name in env) {
    return {
      type: "Identifier",
      name: env[symbol.name].jsVar
    };
  } else {
    return {
      type: "MemberExpression",
      computed: true,
      object: {
        type: "Identifier",
        name: "env"
      },
      property: {
        type: "Literal",
        value: symbol.name
      }
    };
  }
}

export function compile(syntax: ASExpr, env: Environment): JSAST {
  switch (syntax.type) {
    case "number":
      return {
        type: "NumericLiteral",
        value: syntax.value
      };
    case "symbol":
      return compileVariable(syntax, env);
    case "string":
      return {
        type: "Literal",
        value: syntax.value
      };
    case "list":
      const [fn, ...args] = syntax.elements;
      return compileList(fn, args, env);
  }
}

function isSpecialForm(syntax: ASExprList, name: string) {
  if (syntax.elements.length > 0) {
    const first = syntax.elements[0];
    return first.type === "symbol" && first.name === name;
  } else {
    return false;
  }
}

function compileTopLevel(syntax: ASExpr, env: Environment): JSAST {
  if (syntax.type === "list" && isSpecialForm(syntax, "define")) {
    return compileDefinition(syntax.elements.slice(1), env);
  } else {
    return compile(syntax, env);
  }
}

export function compileModule(syntax: ASExpr): JSAST {
  return {
    type: "File",
    program: {
      type: "Program",
      sourceType: "module",
      body: [
        {
          type: "ExpressionStatement",
          expression: compileTopLevel(syntax, {})
        }
      ]
    }
  };
}

export function compileToString(syntax: ASExpr): string {
  const ast = compileModule(syntax);
  debug("jsast:", ast);
  return recast.print(ast).code;
}
