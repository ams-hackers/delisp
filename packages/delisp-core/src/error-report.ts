/**
 * Report errors in a user-friendly way
 */

import { Location } from "./input";

function repeatChar(ch: string, n: number): string {
  return Array(n)
    .fill(ch)
    .join("");
}

/** Print a error message with some highlighted piece of a source code */
export function printHighlightedSource(
  message: string,
  source: string,
  offset: number
) {
  const lines = source.split("\n");

  // Calculate the `line` and `column` (0 based) for this offset in
  // source.
  let line = 0;
  let remainingOffset = offset;
  while (lines.length > line && remainingOffset >= lines[line].length + 1) {
    remainingOffset -= lines[line].length;
    line++;
  }
  const column = remainingOffset;

  return [
    `file:${line + 1}:${column}: ${message}`,
    lines[line],
    repeatChar("-", column) + "^"
  ].join("\n");
}

/** Print a error message with expr highlighted. */
export function printHighlightedExpr(
  message: string,
  location: Location,
  end = false
) {
  const source = location.input.toString();
  const offset = end ? location.end - 1 : location.start;
  return printHighlightedSource(message, source, offset);
}
