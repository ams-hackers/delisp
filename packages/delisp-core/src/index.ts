export * from "./compiler";
export * from "./syntax-convert";
export * from "./error-report";
export * from "./eval";
export * from "./infer";
export * from "./infer-environment";
export * from "./module";
export * from "./module-dependencies";
export * from "./prettier";
export * from "./primitives";
export * from "./printer";
export * from "./syntax";
export * from "./syntax-typed";
export * from "./syntax-scoped";
export * from "./syntax-utils";
export * from "./type-printer";
export * from "./type-utils";
export { Type } from "./types";
export * as T from "./types";
export * from "./typescript-generation";
export * from "./macroexpand";
