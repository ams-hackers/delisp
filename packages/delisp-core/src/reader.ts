//
// reader.ts --- The reader for Delisp
//

// import { List, Syntax } from "./syntax";

//
// S-Expression parser
//

import {
  Parser,
  regex,
  alternatives,
  many,
  delimited
} from "./parser-combinators";

import { ASExpr } from "./syntax";

const spaces = regex(/^\s*/);

function spaced<A>(x: Parser<A>) {
  return delimited(spaces, x, spaces);
}

const number: Parser<ASExpr> = regex(/[0-9]+/)
  .map(
    (string, location): ASExpr => ({
      type: "number",
      value: parseInt(string, 10),
      location
    })
  )
  .description("number");

const doubleQuote = regex(/"/).description("double quote");

const stringChar = regex(/[^"\\]/).description("non-escaped string character");
const stringEscapedChar = regex(/\\[n\\]/).description(
  "escaped string character"
);
const stringConsitutent = alternatives(stringChar, stringEscapedChar);

const string = doubleQuote
  .then(many(stringConsitutent))
  .skip(doubleQuote)
  .map(
    (chars, location): ASExpr => ({
      type: "string",
      value: chars.join(""),
      location
    })
  );

const symbol: Parser<ASExpr> = regex(/[a-zA-Z+<>!@$%^*/-]+/)
  .map(
    (name, location): ASExpr => ({
      type: "symbol",
      name,
      location
    })
  )
  .description("symbol");

const atom: Parser<ASExpr> = alternatives(number, string, symbol).description(
  "atom"
);

const leftParen = regex(/\(/).description("open parenthesis");
const rightParen = regex(/\)/).description("close parenthesis");

function list(x: Parser<ASExpr>): Parser<ASExpr> {
  return leftParen
    .then(many(x))
    .skip(spaces)
    .skip(rightParen)
    .map(
      (elements, location): ASExpr => ({
        type: "list",
        elements,
        location
      })
    );
}

const sexpr: Parser<ASExpr> = spaced(atom.or(() => list(sexpr)));

//
// Parser a Delisp expression from a string
//
export function readFromString(str: string) {
  return sexpr.parse(str);
}
