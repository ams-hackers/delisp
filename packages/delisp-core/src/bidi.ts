import { InvariantViolation, assertNever } from "./invariant";
import { Expression } from "./syntax";
import { Type, tString, tNumber } from "./types";
import { unify } from "./unify";
import { printType } from "./type-printer";
import { last } from "./utils";
import { Location } from "./input";
import { printHighlightedExpr } from "./error-report";

export interface Environment {
  [variable: string]: Type;
}

function equalTypes(t1: Type, t2: Type): boolean {
  // Quick and dirty equality definition for trying out.
  const result = unify(t1, t2, {});
  return (
    result.tag === "unify-success" &&
    Object.keys(result.substitution).length === 0
  );
}

function typeError(message: string, entity: { location: Location }) {
  return new Error(printHighlightedExpr(message, entity.location));
}

export function infer(expr: Expression, env: Environment): Type | null {
  switch (expr.node.tag) {
    case "string":
      return tString;
    case "number":
      return tNumber;
    case "variable-reference": {
      const variableName = expr.node.name;
      const variableType = env[variableName];
      if (!variableType) {
        throw typeError(`Unknown variable ${variableName}`, expr);
      }
      return variableType;
    }
    case "type-annotation": {
      const typ = expr.node.typeWithWildcards.instantiate();
      check(expr.node.value, typ, env);
      return typ;
    }
    case "function-call": {
      const fnType = infer(expr.node.fn, env);
      if (!fnType) {
        throw typeError(
          `Can't called a function with unknown type.`,
          expr.node.fn
        );
      }

      if (
        !(
          fnType.node.tag === "application" &&
          fnType.node.op.node.tag === "constant" &&
          fnType.node.op.node.name === "->"
        )
      ) {
        throw typeError(
          `Looks like you are trying to do a function call on a ${printType(
            fnType
          )}`,
          expr.node.fn
        );
      }

      const argTypes = fnType.node.args.slice(0, -1);
      const outType = last(fnType.node.args);
      if (!outType) {
        throw new InvariantViolation(`Function with no output type?`);
      }

      expr.node.args.forEach((arg, i) => {
        check(arg, argTypes[i], env);
      });

      return outType;
    }

    default:
      return null;
  }
}

export function check(expr: Expression, type: Type, env: Environment) {
  switch (expr.node.tag) {
    case "string":
    case "number":
    case "variable-reference":
    case "type-annotation":
    case "function-call": {
      const inferred = infer(expr, env);
      if (inferred) {
        if (!equalTypes(inferred, type)) {
          throw typeError(
            `Expected type ${printType(type, false)}, but found ${printType(
              inferred
            )}.`,
            expr
          );
        }
      }
      return;
    }
    case "function": {
      if (
        !(
          type.node.tag === "application" &&
          type.node.op.node.tag === "constant" &&
          type.node.op.node.name === "->"
        )
      ) {
        throw typeError(
          `Expected a function, but found ${printType(type)}`,
          expr
        );
      }

      const argTypes = type.node.args.slice(0, -1);
      const outType = last(type.node.args);
      if (!outType) {
        throw new InvariantViolation(`Function with no output type?`);
      }

      const args = expr.node.lambdaList.positionalArgs.map(i => i.name);

      if (argTypes.length !== args.length) {
        throw typeError(
          `Wrong number of arguments. Expected ${argTypes.length} but found ${
            args.length
          }`,
          expr.node.lambdaList
        );
      }

      const newEnv = args.reduce(
        (tmpEnv, a, idx) => ({
          ...tmpEnv,
          [a]: argTypes[idx]
        }),
        env
      );

      check(last(expr.node.body)!, outType, newEnv);
      return;
    }
    case "conditional":
    case "let-bindings":
    case "record":
    case "vector":
      throw typeError(
        `Cannot check this expression for type ${printType(type)}.`,
        expr
      );
    default:
      assertNever(expr.node);
  }
}
