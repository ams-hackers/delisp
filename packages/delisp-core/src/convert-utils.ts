import { printHighlightedExpr } from "./error-report";
import { ASExpr, ASExprMap, ASExprSymbol } from "./sexpr";
import { duplicatedItemsBy, last } from "./utils";

export class ConvertError extends Error {}

export interface ParseRecordResult {
  fields: Array<{
    label: ASExprSymbol;
    value: ASExpr;
  }>;
  tail: ASExpr | undefined;
}

export function parseRecord(expr: ASExprMap): ParseRecordResult {
  //
  // Destructure a map into fields and tail
  function fieldsAndTail() {
    if (expr.fields.length === 0) {
      return { fields: expr.fields };
    } else {
      const lastField = last(expr.fields)!;
      return lastField.label.name === "|"
        ? {
            fields: expr.fields.slice(0, -1),
            tail: lastField.value
          }
        : { fields: expr.fields };
    }
  }

  function checkInvalidFields() {
    const invalidFields = fields.filter(f => !f.label.name.startsWith(":"));
    if (invalidFields.length > 0) {
      const field = invalidFields[0];
      throw new ConvertError(
        printHighlightedExpr(
          `'${
            field.label.name
          }' is not a valid field name. Fields should start with ':' (colon).`,
          field.label.location
        )
      );
    }
  }

  const { fields, tail } = fieldsAndTail();
  checkInvalidFields();

  const duplicates = duplicatedItemsBy(fields, f => f.label.name);
  if (duplicates.length > 0) {
    throw new ConvertError(
      printHighlightedExpr("Duplicated label", duplicates[0].label.location)
    );
  }

  return { fields, tail };
}
