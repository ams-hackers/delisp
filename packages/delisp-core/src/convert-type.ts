import { parseRecord } from "./convert-utils";
import { printHighlightedExpr } from "./error-report";
import {
  ASExpr,
  ASExprList,
  ASExprMap,
  ASExprSymbol,
  ASExprVector
} from "./sexpr";
import { generateUniqueTVar } from "./type-generate";
import {
  emptyRow,
  Monotype,
  tApp,
  tBoolean,
  tNumber,
  tRecord,
  tString,
  tUserDefined,
  tVar,
  tVector,
  tVoid
} from "./types";

/** Capitalize a string like "foo" to "Foo". */
function capitalize(str: string) {
  if (str === "") {
    return "";
  } else {
    return str[0].toUpperCase() + str.slice(1);
  }
}

/** Return true if a symbol is a valid name for a user defined type, false otherwise. */
export function userDefinedType(expr: ASExprSymbol): boolean {
  return expr.name[0] === expr.name[0].toUpperCase();
}

/** Check if a symbol is a valid user defined type name or throw a user-friendly error otherwise. */
export function checkUserDefinedTypeName(expr: ASExprSymbol): void {
  if (!userDefinedType(expr)) {
    throw new Error(
      printHighlightedExpr(
        `'${
          expr.name
        }' is not a valid name for a type. Type names should start with a capital letter. Try '${capitalize(
          expr.name
        )}'?`,
        expr.location
      )
    );
  }
}

function convertSymbol(expr: ASExprSymbol): Monotype {
  switch (expr.name) {
    case "boolean":
      return tBoolean;
    case "number":
      return tNumber;
    case "string":
      return tString;
    case "void":
      return tVoid;
    case "_":
      return generateUniqueTVar(false, "__t");
    default:
      return userDefinedType(expr) ? tUserDefined(expr.name) : tVar(expr.name);
  }
}

function convertList(expr: ASExprList): Monotype {
  const [op, ...args] = expr.elements;

  if (op.type !== "symbol") {
    throw new Error(
      printHighlightedExpr("Expected symbol as operator", expr.location)
    );
  }

  switch (op.name) {
    case "->":
      if (args.length < 1) {
        throw new Error(
          printHighlightedExpr(
            "Expected at least 1 argument",
            op.location,
            true
          )
        );
      }
      break;
    default:
      throw new Error(
        printHighlightedExpr("Unknown type constructor", op.location)
      );
  }

  return tApp(op.name, ...args.map(convert));
}

function convertVector(expr: ASExprVector): Monotype {
  if (expr.elements.length !== 1) {
    throw new Error(
      printHighlightedExpr("Expected exactly 1 argument", expr.location)
    );
  }
  return tVector(convert(expr.elements[0]));
}

function convertMap(expr: ASExprMap): Monotype {
  const { fields, tail } = parseRecord(expr);

  return tRecord(
    fields.map(({ label, value }) => ({
      label: label.name,
      type: convert(value)
    })),
    tail ? convert(tail) : emptyRow
  );
}

/* Try to convert a S-Expression into a type. */
export function convert(expr: ASExpr): Monotype {
  switch (expr.type) {
    case "list":
      return convertList(expr);
    case "symbol":
      return convertSymbol(expr);
    case "vector":
      return convertVector(expr);
    case "map":
      return convertMap(expr);
    default:
      throw new Error(printHighlightedExpr("Not a valid type", expr.location));
  }
}
