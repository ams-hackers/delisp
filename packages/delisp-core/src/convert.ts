import { assertNever } from "./invariant";
import { Location } from "./input";
import { printHighlightedExpr } from "./error-report";
import {
  ASExpr,
  ASExprList,
  ASExprMap,
  ASExprSymbol,
  ASExprVector
} from "./sexpr";
import {
  isExpression,
  Identifier,
  Declaration,
  Expression,
  SLetBindingF,
  LambdaList,
  Syntax
} from "./syntax";
import { foldExpr, exprFChildren } from "./syntax-utils";
import { flatten, last } from "./utils";

import {
  checkUserDefinedTypeName,
  convert as convertType
} from "./convert-type";
import { ConvertError, parseRecord } from "./convert-utils";
import { listTypeVariables } from "./type-utils";
import { TypeWithWildcards } from "./type-wildcards";

export interface WithErrors {
  errors: string[];
}

type ExpressionWithErrors = Expression<WithErrors>;
type DeclarationWithErrors = Declaration<WithErrors, WithErrors>;
type SyntaxWithErrors = Syntax<WithErrors, WithErrors>;

function result(
  node: ExpressionWithErrors["node"],
  location: Location,
  errors: string[]
): ExpressionWithErrors {
  return { node, info: { errors }, location };
}

function success(
  node: ExpressionWithErrors["node"],
  location: Location
): ExpressionWithErrors {
  return result(node, location, []);
}

const conversions: Map<
  string,
  (expr: ASExprList) => ExpressionWithErrors
> = new Map();
const toplevelConversions: Map<
  string,
  (expr: ASExprList) => DeclarationWithErrors
> = new Map();

function defineConversion(
  name: string,
  fn: (expr: ASExprList) => ExpressionWithErrors
) {
  conversions.set(name, fn);
}

function defineToplevel(
  name: string,
  fn: (expr: ASExprList) => DeclarationWithErrors
) {
  toplevelConversions.set(name, fn);
}

function parseBody(anchor: ASExpr, exprs: ASExpr[]): ExpressionWithErrors[] {
  if (exprs.length === 0) {
    throw new ConvertError(
      printHighlightedExpr(`body can't be empty`, anchor.location, true)
    );
  }
  return exprs.map(convertExpr);
}

//
// The format of lambda lists are (a b c ... &rest z)
//

function parseLambdaList(ll: ASExpr): LambdaList {
  if (ll.tag !== "list") {
    throw new ConvertError(
      printHighlightedExpr("Expected a list of arguments", ll.location)
    );
  }

  ll.elements.forEach(arg => {
    if (arg.tag !== "symbol") {
      throw new ConvertError(
        printHighlightedExpr(
          "A list of arguments should be made of symbols",
          arg.location
        )
      );
    }
  });

  const symbols = ll.elements as ASExprSymbol[];

  // Check for duplicated arguments
  symbols.forEach((arg, i) => {
    const duplicated = symbols.slice(i + 1).find(a => a.name === arg.name);
    if (duplicated) {
      throw new ConvertError(
        printHighlightedExpr(
          "There is another argument with the same name",
          duplicated.location
        )
      );
    }
  });

  return {
    positionalArgs: symbols.map(parseIdentifier),
    location: ll.location
  };
}

defineConversion("lambda", expr => {
  const [lambda, ...args] = expr.elements;
  const lastExpr = last([lambda, ...args]) as ASExpr; // we kj

  if (args.length === 0) {
    throw new ConvertError(
      printHighlightedExpr(
        `'lambda' is missing the argument list`,
        lastExpr.location,
        true
      )
    );
  }

  const body = parseBody(lastExpr, args.slice(1));

  return success(
    {
      tag: "function",
      lambdaList: parseLambdaList(args[0]),
      body
    },
    expr.location
  );
});

function missingFrom(expr: ASExpr): ExpressionWithErrors {
  const location = {
    ...expr.location,
    start: expr.location.end,
    end: expr.location.end
  };
  return success({ tag: "unknown" }, location);
}

function exactArguments(
  args: ASExpr[],
  options: {
    required: number;
    fewArguments: string;
    manyArguments: string;
  }
): string[] {
  if (args.length < options.required) {
    return [
      printHighlightedExpr(options.fewArguments, last(args)!.location, true)
    ];
  }
  if (args.length > options.required) {
    return [
      printHighlightedExpr(
        options.fewArguments,
        args[options.required].location,
        true
      )
    ];
  }
  return [];
}

defineConversion("if", expr => {
  const [, conditionForm, consequentForm, alternativeForm] = expr.elements;

  const condition = conditionForm
    ? convertExpr(conditionForm)
    : missingFrom(expr);
  const consequent = consequentForm
    ? convertExpr(consequentForm)
    : missingFrom(expr);
  const alternative = alternativeForm
    ? convertExpr(alternativeForm)
    : missingFrom(expr);

  const receivedArgs = expr.elements.length - 1;

  const errors = exactArguments(expr.elements, {
    required: 4,
    fewArguments: `'if' needs exactly 3 arguments, got ${receivedArgs}`,
    manyArguments: `'if' needs exactly 3 arguments, got ${receivedArgs}`
  });

  return result(
    {
      tag: "conditional",
      condition,
      consequent,
      alternative
    },
    expr.location,
    errors
  );
});

function parseLetBindings(
  bindings: ASExpr
): Array<SLetBindingF<ExpressionWithErrors>> {
  if (bindings.tag !== "map") {
    throw new ConvertError(
      printHighlightedExpr(`'let' bindings should be a map`, bindings.location)
    );
  }
  return bindings.fields.map(field => ({
    variable: parseIdentifier(field.label),
    value: convertExpr(field.value)
  }));
}

defineConversion("let", expr => {
  const [_let, ...args] = expr.elements;
  const lastExpr = last([_let, ...args]) as ASExpr; // we know it is not empty!

  if (args.length === 0) {
    throw new ConvertError(
      printHighlightedExpr(
        `'let' is missing the bindings`,
        lastExpr.location,
        true
      )
    );
  }

  const [rawBindings, ...rawBody] = args;

  return success(
    {
      tag: "let-bindings",
      bindings: parseLetBindings(rawBindings),
      body: parseBody(lastExpr, rawBody)
    },
    expr.location
  );
});

defineConversion("the", expr => {
  const [_the, ...args] = expr.elements;
  const lastExpr = last([_the, ...args]) as ASExpr; // we know it is not empty!

  if (args.length === 0) {
    throw new ConvertError(
      printHighlightedExpr(
        `'the' is missing the type and value`,
        lastExpr.location,
        true
      )
    );
  }
  if (args.length === 1) {
    throw new ConvertError(
      printHighlightedExpr(
        `'the' is missing the expression`,
        lastExpr.location,
        true
      )
    );
  }

  if (args.length > 2) {
    throw new ConvertError(
      printHighlightedExpr(
        `Too many arguments. 'the' should take two arguments.`,
        args[2].location,
        true
      )
    );
  }

  const [t, value] = args;

  return success(
    {
      tag: "type-annotation",
      typeWithWildcards: new TypeWithWildcards(convertType(t)),
      value: convertExpr(value)
    },
    expr.location
  );
});

defineConversion("do", expr => {
  const [_do, ...args] = expr.elements;
  const lastExpr = last([_do, ...args]) as ASExpr; // we know it is not empty!

  if (args.length === 0) {
    throw new ConvertError(
      printHighlightedExpr(`empty body`, lastExpr.location, true)
    );
  }
  const middleForms = args.slice(0, -1);
  const lastForm = last(args)!;
  return success(
    {
      tag: "do-block",
      body: middleForms.map(convertExpr),
      returning: convertExpr(lastForm)
    },
    expr.location
  );
});

defineToplevel("define", expr => {
  const [define, ...args] = expr.elements;

  if (args.length !== 2) {
    const lastExpr = last([define, ...args]) as ASExpr;
    throw new ConvertError(
      printHighlightedExpr(
        `'define' needs exactly 2 arguments, got ${args.length}`,
        lastExpr.location,
        true
      )
    );
  }

  const [variable, value] = args;

  if (variable.tag !== "symbol") {
    throw new ConvertError(
      printHighlightedExpr("'define' expected a symbol", variable.location)
    );
  }

  return {
    node: {
      tag: "definition",
      variable: parseIdentifier(variable),
      value: convertExpr(value)
    },
    location: expr.location,
    info: {
      errors: []
    }
  };
});

defineToplevel("export", expr => {
  const [exp, ...args] = expr.elements;

  if (args.length !== 1) {
    const lastExpr = last([exp, ...args]) as ASExpr;
    throw new ConvertError(
      printHighlightedExpr(
        `'export' needs exactly 1 arguments, got ${args.length}`,
        lastExpr.location,
        true
      )
    );
  }

  const [variable] = args;

  if (variable.tag !== "symbol") {
    throw new ConvertError(
      printHighlightedExpr("'export' expected a symbol", variable.location)
    );
  }

  return {
    node: {
      tag: "export",
      value: parseIdentifier(variable)
    },
    location: expr.location,
    info: {
      errors: []
    }
  };
});

defineToplevel("type", expr => {
  const [typeOp, ...args] = expr.elements;

  if (args.length !== 2) {
    const lastExpr = last([typeOp, ...args]) as ASExpr;
    throw new ConvertError(
      printHighlightedExpr(
        `'type' needs exactly 2 arguments, got ${args.length}`,
        lastExpr.location,
        true
      )
    );
  }

  const [name, definition] = args;

  if (name.tag !== "symbol") {
    throw new ConvertError(
      printHighlightedExpr("'type' expected a symbol as a name", name.location)
    );
  }

  checkUserDefinedTypeName(name);

  const definitionType = convertType(definition);
  if (listTypeVariables(definitionType).length > 0) {
    throw new ConvertError(
      printHighlightedExpr("Type variable out of scope", definition.location)
    );
  }

  return {
    node: {
      tag: "type-alias",
      alias: parseIdentifier(name),
      definition: definitionType
    },
    location: expr.location,
    info: {
      errors: []
    }
  };
});

function convertList(list: ASExprList): ExpressionWithErrors {
  if (list.elements.length === 0) {
    throw new ConvertError(
      printHighlightedExpr("Empty list is not a function call", list.location)
    );
  }

  const [first] = list.elements;

  const convertSpecialForm =
    first.tag === "symbol" ? conversions.get(first.name) : undefined;

  if (convertSpecialForm) {
    return convertSpecialForm(list);
  } else {
    const [fn, ...args] = list.elements;
    return success(
      {
        tag: "function-call",
        fn: convertExpr(fn),
        args: args.map(convertExpr)
      },
      list.location
    );
  }
}

function convertVector(list: ASExprVector): ExpressionWithErrors {
  return success(
    {
      tag: "vector",
      values: list.elements.map(convertExpr)
    },
    list.location
  );
}

function convertMap(map: ASExprMap): ExpressionWithErrors {
  const { fields, tail } = parseRecord(map);
  return success(
    {
      tag: "record",
      fields: fields.map(f => ({
        label: parseIdentifier(f.label),
        value: convertExpr(f.value)
      })),
      extends: tail && convertExpr(tail)
    },
    map.location
  );
}

function parseIdentifier(expr: ASExprSymbol): Identifier {
  return {
    tag: "identifier",
    name: expr.name,
    location: expr.location
  };
}

function convertSymbol(expr: ASExprSymbol): ExpressionWithErrors {
  const id = parseIdentifier(expr);
  return success(
    {
      tag: "variable-reference",
      name: id.name
    },
    id.location
  );
}

function convertExprOrError(expr: ASExpr): ExpressionWithErrors {
  switch (expr.tag) {
    case "number":
      return success({ tag: "number", value: expr.value }, expr.location);
    case "string":
      return success({ tag: "string", value: expr.value }, expr.location);
    case "symbol":
      return convertSymbol(expr);
    case "list":
      return convertList(expr);
    case "vector":
      return convertVector(expr);
    case "map":
      return convertMap(expr);
  }
}

function convertExpr(expr: ASExpr): ExpressionWithErrors {
  try {
    return convertExprOrError(expr);
  } catch (err) {
    if (err instanceof ConvertError) {
      return result({ tag: "unknown" }, expr.location, [err.message]);
    } else {
      throw err;
    }
  }
}

function convertOrError(form: ASExpr): SyntaxWithErrors {
  if (form.tag === "list") {
    if (form.elements.length === 0) {
      throw new ConvertError(
        printHighlightedExpr("Empty list is not a function call", form.location)
      );
    }

    const [first] = form.elements;
    const convertDeclaration =
      first.tag === "symbol" ? toplevelConversions.get(first.name) : undefined;

    if (convertDeclaration) {
      return convertDeclaration(form);
    }
  }

  return convertExpr(form);
}

export function convert(form: ASExpr): SyntaxWithErrors {
  try {
    return convertOrError(form);
  } catch (err) {
    if (err instanceof ConvertError) {
      return result({ tag: "unknown" }, form.location, [err.message]);
    } else {
      throw err;
    }
  }
}

//
// Extracing errors
//

function collectConvertExprErrors(expr: ExpressionWithErrors): string[] {
  return foldExpr(expr, e => [...e.info.errors, ...flatten(exprFChildren(e))]);
}

export function collectConvertErrors(syntax: SyntaxWithErrors): string[] {
  if (isExpression(syntax)) {
    return collectConvertExprErrors(syntax);
  } else {
    switch (syntax.node.tag) {
      case "export":
      case "type-alias":
        return syntax.info.errors;
      case "definition":
        return [
          ...syntax.info.errors,
          ...collectConvertExprErrors(syntax.node.value)
        ];
      default:
        return assertNever(syntax.node);
    }
  }
}
