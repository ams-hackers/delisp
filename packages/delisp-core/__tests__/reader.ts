import { readFromString } from "../src/reader";

describe("Reader", () => {
  it("should read basic booleans", () => {
    expect(readFromString("(1 2 3)")).toEqual({
      type: "LiteralBoolean",
      value: true
    });
  });

  // it.skip("should read basic expessions", () => {
  //   expect(readFromString("1")).toEqual({ type: "LiteralNumber", value: 1 });
  //   expect(readFromString("x")).toEqual({ type: "LiteralSymbol", value: "x" });
  //   expect(readFromString("()")).toEqual({ type: "List", value: [] });
  //   expect(readFromString("(1 2)")).toEqual({
  //     type: "List",
  //     value: [
  //       { type: "LiteralNumber", value: 1 },
  //       { type: "LiteralNumber", value: 2 }
  //     ]
  //   });
  //   // expect(readFromString("(whatever (stuff)"))
  // });
  // it.skip("should parse nested lists", () => {
  //   expect(readFromString(" ( 1 ( 2 ) 3 )")).toEqual({
  //     type: "List",
  //     value: [
  //       { type: "LiteralNumber", value: 1 },
  //       {
  //         type: "List",
  //         value: [{ type: "LiteralNumber", value: 2 }]
  //       },
  //       { type: "LiteralNumber", value: 3 }
  //     ]
  //   });
  // });
});
