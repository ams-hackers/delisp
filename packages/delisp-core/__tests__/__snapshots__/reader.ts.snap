// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reader should generate user-friendly errors 1`] = `
"file:1:6: Expected close parenthesis
(1 2 3
------^"
`;

exports[`Reader should generate user-friendly errors 2`] = `
"file:1:0: Expected Unmatched closed parenthesis
)
^"
`;

exports[`Reader should generate user-friendly errors 3`] = `
"file:1:4: Expected double quote
\\"foo
----^"
`;

exports[`Reader should generate user-friendly errors 4`] = `
"file:1:3: Expected double quote
\\"ab\\\\xyz\\"
---^"
`;
